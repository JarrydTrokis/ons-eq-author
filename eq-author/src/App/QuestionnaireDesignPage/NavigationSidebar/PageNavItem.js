import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { withRouter } from "react-router-dom";
import CustomPropTypes from "custom-prop-types";
import gql from "graphql-tag";
import { flowRight } from "lodash";

import { buildPagePath } from "utils/UrlUtils";
import { withValidations } from "components/ValidationsContext";

import NavLink from "./NavLink";
import PageIcon from "./icon-questionpage.svg?inline";
import CalculatedIcon from "./icon-summarypage.svg?inline";

const StyledPageItem = styled.li`
  padding: 0;
  margin: 0;
  position: relative;
  display: flex;
  align-items: center;
`;

const getIcon = pageType => {
  if (pageType === "CalculatedSummaryPage") {
    return CalculatedIcon;
  }
  return PageIcon;
};

export const UnwrappedPageNavItem = ({
  questionnaireId,
  page,
  match,
  validations,
  ...otherProps
}) => {
  let errorCount = page.validationErrorInfo.totalCount;
  if (validations) {
    const pageErrors = validations.pages.find(
      validations => validations.id === page.id
    );
    errorCount = pageErrors ? pageErrors.errorCount : 0;
  }
  return (
    <StyledPageItem data-test="page-item" {...otherProps}>
      <NavLink
        to={buildPagePath({
          questionnaireId,
          pageId: page.id,
          tab: match.params.tab,
        })}
        title={page.displayName}
        icon={getIcon(page.pageType)}
        data-test="nav-page-link"
        errorCount={errorCount}
      >
        {page.displayName}
      </NavLink>
    </StyledPageItem>
  );
};

UnwrappedPageNavItem.fragments = {
  PageNavItem: gql`
    fragment PageNavItem on Page {
      id
      title
      position
      displayName
      pageType

      ... on QuestionPage {
        confirmation {
          id
        }
        validationErrorInfo {
          totalCount
        }
      }
    }
  `,
};

UnwrappedPageNavItem.propTypes = {
  questionnaireId: PropTypes.string.isRequired,
  page: CustomPropTypes.page,
  match: CustomPropTypes.match,
  validations: PropTypes.shape({
    pages: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.string.isRequired,
        errorCount: PropTypes.number.isRequired,
      }).isRequired
    ).isRequired,
  }),
};

export default flowRight(
  withRouter,
  withValidations
)(UnwrappedPageNavItem);
