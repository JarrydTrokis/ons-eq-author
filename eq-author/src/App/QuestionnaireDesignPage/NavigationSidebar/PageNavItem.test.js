import React from "react";
import { shallow } from "enzyme";

import NavLink from "./NavLink";
import { UnwrappedPageNavItem } from "./PageNavItem";

describe("PageNavItem", () => {
  let props;
  beforeEach(() => {
    const page = {
      id: "3",
      displayName: "Page title",
      validationErrorInfo: {
        totalCount: 0,
      },
    };

    const match = {
      params: {
        questionnaireId: "1",
        sectionId: "2",
        pageId: page.id,
      },
    };

    props = {
      questionnaireId: "1",
      sectionId: "2",
      title: "Title",
      page,
      match,
    };
  });

  it("should render", () => {
    const wrapper = shallow(<UnwrappedPageNavItem {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should show the original load errors when no validations from context", () => {
    props.page.validationErrorInfo = {
      totalCount: 2,
    };
    props.validations = null;
    const wrapper = shallow(<UnwrappedPageNavItem {...props} />);
    expect(wrapper.find(NavLink).props()).toMatchObject({
      errorCount: 2,
    });
  });

  it("should show the context validation errors it has a context", () => {
    props.page.validationErrorInfo = {
      totalCount: 2,
    };
    props.validations = {
      pages: [{ id: props.page.id, errorCount: 1 }],
    };
    const wrapper = shallow(<UnwrappedPageNavItem {...props} />);
    expect(wrapper.find(NavLink).props()).toMatchObject({
      errorCount: 1,
    });
  });

  it("should show no errors when the context has validations but not for this page", () => {
    props.page.validationErrorInfo = {
      totalCount: 2,
    };
    props.validations = {
      pages: [],
    };
    const wrapper = shallow(<UnwrappedPageNavItem {...props} />);
    expect(wrapper.find(NavLink).props()).toMatchObject({
      errorCount: 0,
    });
  });
});
