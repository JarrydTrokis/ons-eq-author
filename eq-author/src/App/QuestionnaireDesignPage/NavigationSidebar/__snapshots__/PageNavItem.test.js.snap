// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageNavItem should render 1`] = `
<PageNavItem__StyledPageItem
  data-test="page-item"
  sectionId="2"
  title="Title"
>
  <NavLink
    data-test="nav-page-link"
    icon={[Function]}
    title="Page title"
    to="/q/1/page/3/design"
  >
    Page title
  </NavLink>
</PageNavItem__StyledPageItem>
`;
